### PR TITLE
Restrict configured indices access to adminDn only.

### PR DESCRIPTION
**Pull request by skkosuri-amzn**
*Issue #, if available:*
https://github.com/opendistro-for-elasticsearch/security/issues/666

*Description of changes:*
https://github.com/opendistro-for-elasticsearch/security/issues/666

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: src/test/java/com/amazon/opendistroforelasticsearch/security/system_indices/SystemIndicesTests.java [hunk 1]
```diff
@@ -0,0 +1,547 @@
...
+
+
+    @Test
+    public void testSearchWithSystemIndicesAsAdmin() throws Exception {
+        setupSystemIndicesEnabledWithSsl();
+        createTestIndicesAndDocs();
+        RestHelper restHelper = sslRestHelper();
+
+        for (String index : listOfIndexesToTest) {
+            validateSearchResponse(restHelper.executePostRequest(index + "/_search", matchAllQuery, allAccessUserHeader), 0);
```
Comment (hardik-k-shah) [inline src/test/java/com/amazon/opendistroforelasticsearch/security/system_indices/SystemIndicesTests.java:210]: why are we allowing admin to search system indices? 

System indices will be accessible through es APIs only for AdminDN.  All the other user including admin will be able to access index only through APIs defined by respective plugin.

File: src/main/java/com/amazon/opendistroforelasticsearch/security/privileges/OpenDistroSystemIndexAccessEvaluator.java [hunk 2]
```diff
@@ -0,0 +1,103 @@
...
+    private final WildcardMatcher deniedActionMatcher;
+
+
+    public OpenDistroSystemIndexAccessEvaluator(final Settings settings, AuditLog auditLog) {
+        this.indexMatcher = WildcardMatcher.from(settings.getAsList(ConfigConstants.OPENDISTRO_SECURITY_SYSTEM_INDICES_KEY, ConfigConstants.OPENDISTRO_SECURITY_SYSTEM_INDICES_DEFAULT));
+        this.systemIndexEnabled = settings.getAsBoolean(ConfigConstants.OPENDISTRO_SECURITY_SYSTEM_INDICES_ENABLED_KEY, ConfigConstants.OPENDISTRO_SECURITY_SYSTEM_INDICES_ENABLED_DEFAULT);
+        this.auditLog = auditLog;
+
+        final List<String> indexDeniedActionPatterns = new ArrayList<String>();
+        indexDeniedActionPatterns.add("indices:data/write*");
```
[the remark this answers is not in the document]
Comment (skkosuri-amzn) [reply src/main/java/com/amazon/opendistroforelasticsearch/security/privileges/OpenDistroSystemIndexAccessEvaluator.java:52]: Followed the this discussion : 1/ https://github.com/opendistro-for-elasticsearch/security/pull/126#discussion_r320539005
2/ https://github.com/opendistro-for-elasticsearch/security/pull/126#discussion_r320539568

[the remark this answers is not in the document]
Comment (skkosuri-amzn) [reply src/main/java/com/amazon/opendistroforelasticsearch/security/privileges/OpenDistroSystemIndexAccessEvaluator.java:52]: Followed below 3 links
https://github.com/opendistro-for-elasticsearch/security/pull/126#discussion_r320539005
https://github.com/opendistro-for-elasticsearch/security/pull/126#discussion_r320539568
https://github.com/opendistro-for-elasticsearch/security/blob/d62ac5334609ea92b913caa5e73917efb8b29043/src/main/java/com/amazon/opendistroforelasticsearch/security/privileges/OpenDistroSecurityIndexAccessEvaluator.java#L71-L82

OpenDistroSecurityIndexSearcherWrapper blocks indices:data/write*. indices:data/read* is blocked here in OpenDistroSystemIndexAccessEvaluator.

On indices:admin/*, I tried to make this behavior is same as security index. (OpenDistroSecurityIndexAccessEvaluator).

File: src/main/java/com/amazon/opendistroforelasticsearch/security/privileges/OpenDistroSystemIndexAccessEvaluator.java [hunk 3]
```diff
@@ -0,0 +1,103 @@
...
+            log.warn(action + " for '{}' index/indices is not allowed for a non adminDN user", indexMatcher);
+            presponse.allowed = false;
+            return presponse.markComplete();
+        }
+
+        if (requestedResolved.isLocalAll() && deniedActionMatcher.test(action) && !isAdminDn) {
+            auditLog.logMissingPrivileges(action, request, task);
+            log.warn(action + " for '_all' indices is not allowed for a non adminDN user");
+            presponse.allowed = false;
+            return presponse.markComplete();
```
[the remark this answers is not in the document]
Comment (skkosuri-amzn) [reply src/main/java/com/amazon/opendistroforelasticsearch/security/privileges/OpenDistroSystemIndexAccessEvaluator.java:82]: Good point. "_all" is currently caught by OpenDistroSecurityIndexAccessEvaluator. I don't see a case (with _all) which will pass OpenDistroSecurityIndexAccessEvaluator and reaches here. I am tending to remove this check with isLocalAll(). What do you think?

File: src/test/java/com/amazon/opendistroforelasticsearch/security/system_indices/SystemIndicesTests.java (hunk 1 above)
Comment (skkosuri-amzn) [reply src/test/java/com/amazon/opendistroforelasticsearch/security/system_indices/SystemIndicesTests.java:210]: No, we are not allowing `admin` to search system indices. `admin` requests return 200 but with empty results. This is same behavior as security index. This allows `_all\_search` request work, but filters results from system indices.

Comment (hardik-k-shah) [reply src/test/java/com/amazon/opendistroforelasticsearch/security/system_indices/SystemIndicesTests.java:210]: Thanks for clarifying!

File: src/main/java/com/amazon/opendistroforelasticsearch/security/privileges/OpenDistroSystemIndexAccessEvaluator.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (skkosuri-amzn) [reply src/main/java/com/amazon/opendistroforelasticsearch/security/privileges/OpenDistroSystemIndexAccessEvaluator.java]: Had a offline discussion with @hardik-k-shah and we decided to merge this logic into OpenDistroSecurityIndexAccessEvaluator

**Comment by hardik-k-shah** [comment]
Thanks for making optimizations as discussed and removing additional handler for system indices. Merging system indices access check with security index access handler will help us not to introduce any performance penalties and also help to avoid code duplicacy.

File: src/main/java/com/amazon/opendistroforelasticsearch/security/configuration/OpenDistroSecurityIndexSearcherWrapper.java [hunk 4]
```diff
@@ -87,7 +93,12 @@ public final DirectoryReader apply(DirectoryReader reader) throws IOException {
         if (isSecurityIndexRequest() && !isAdminAuthenticatedOrInternalRequest()) {
```
Comment (hardik-k-shah) [inline src/main/java/com/amazon/opendistroforelasticsearch/security/configuration/OpenDistroSecurityIndexSearcherWrapper.java:93]: should we add isSystemIndexRequest() check here?

File: src/main/java/com/amazon/opendistroforelasticsearch/security/privileges/PrivilegesEvaluator.java [hunk 5]
```diff
@@ -36,7 +36,6 @@
 import java.util.Iterator;
 import java.util.List;
 import java.util.Map;
-import java.util.Objects;
```
Comment (hardik-k-shah) [inline src/main/java/com/amazon/opendistroforelasticsearch/security/privileges/PrivilegesEvaluator.java:39]: make sure this is not by mistake..

File: src/test/java/com/amazon/opendistroforelasticsearch/security/system_indices/SystemIndicesTests.java [hunk 6]
```diff
@@ -0,0 +1,547 @@
+/*
+ * Portions Copyright 2019 Amazon.com, Inc. or its affiliates. All Rights Reserved.
```
Comment (hardik-k-shah) [inline src/test/java/com/amazon/opendistroforelasticsearch/security/system_indices/SystemIndicesTests.java:2]: "Portions" is not required. Amazon has full copyright on this file.

File: src/main/java/com/amazon/opendistroforelasticsearch/security/configuration/OpenDistroSecurityIndexSearcherWrapper.java (hunk 4 above)
Comment (skkosuri-amzn) [reply src/main/java/com/amazon/opendistroforelasticsearch/security/configuration/OpenDistroSecurityIndexSearcherWrapper.java:93]: Had a offline discussion and decided to keep it current way.

Comment (vrozov) [reply src/main/java/com/amazon/opendistroforelasticsearch/security/configuration/OpenDistroSecurityIndexSearcherWrapper.java:93]: IMO, the entire class needs to be modified. I'll do this in a follow up PR.

File: src/main/java/com/amazon/opendistroforelasticsearch/security/privileges/OpenDistroSecurityIndexAccessEvaluator.java [hunk 7]
```diff
@@ -58,12 +58,18 @@
     private final WildcardMatcher securityDeniedActionMatcher;
     private final IndexResolverReplacer irr;
     private final boolean filterSecurityIndex;
-    
+
+    // for system-indices configuration
+    private final WildcardMatcher systemIndexMatcher;
+    private final Boolean systemIndexEnabled;
```
Comment (vrozov) [inline src/main/java/com/amazon/opendistroforelasticsearch/security/privileges/OpenDistroSecurityIndexAccessEvaluator.java:64]: `Boolean` -> `boolean`

File: src/main/java/com/amazon/opendistroforelasticsearch/security/privileges/OpenDistroSecurityIndexAccessEvaluator.java [hunk 8]
```diff
@@ -147,4 +153,8 @@ public PrivilegesEvaluatorResponse evaluate(final ActionRequest request, final T
         }
         return presponse;
     }
+
+    private boolean matchAnySystemIndices(final Resolved requestedResolved){
+        return systemIndexEnabled & systemIndexMatcher.matchAny(requestedResolved.getAllIndices());
```
Comment (vrozov) [inline src/main/java/com/amazon/opendistroforelasticsearch/security/privileges/OpenDistroSecurityIndexAccessEvaluator.java:158]: use Java logical `&&`, not bitwise `&`.

Comment (skkosuri-amzn) [reply src/main/java/com/amazon/opendistroforelasticsearch/security/privileges/OpenDistroSecurityIndexAccessEvaluator.java:158]: 👍

File: src/main/java/com/amazon/opendistroforelasticsearch/security/privileges/OpenDistroSecurityIndexAccessEvaluator.java (hunk 7 above)
Comment (skkosuri-amzn) [reply src/main/java/com/amazon/opendistroforelasticsearch/security/privileges/OpenDistroSecurityIndexAccessEvaluator.java:64]: changed

**Review by vrozov — APPROVED** [review]

File: securityconfig/elasticsearch.yml.example [hunk 9]
```diff
@@ -217,3 +217,9 @@ opendistro_security.audit.type: internal_elasticsearch
 # Specify a list of indices to mark as protected. These indices will only be visible / mutable by members of the above setting, in addition to needing permission to the index via a normal role.
 # opendistro_security.protected_indices.indices: ['.opendistro-alerting-config', '.opendistro-ism-*']
 
+# System indices are similar to security index, except the contents are not encrypted.
+# Indices configured as system indices can be accessed by only super-admin and no role will provide access to these indices.
+# Enable system indices
+# opendistro_security.system_indices.enabled: true
```
[the remark this answers is not in the document]
Comment (skkosuri-amzn) [reply securityconfig/elasticsearch.yml.example:223]: @hardik-k-shah : we are planning that a separate PR, once other plugins are also ready.

[the remark this answers is not in the document]
Comment (hardik-k-shah) [reply securityconfig/elasticsearch.yml.example:223]: Okay

**Review by hardik-k-shah — APPROVED** [review]

**Pull request merged**